src/reducers/versions.tsx

```diff
@@ -1138,9 +1150,10 @@ export const goToRelativeDiff = ({
   };
 };
 
-const reducer: Reducer<VersionsState, ActionType<typeof actions>> = (
```

Comment:
Ah, whoops. I forgot that you need to define it as a function that returns a reducer. I can use the *hide whitespace* feature in github to review the patch since it will indent a lot of code.

<details>
<summary>Example patch</summary>

```diff
diff --git a/src/reducers/versions.spec.tsx b/src/reducers/versions.spec.tsx
index 2bc92d9..a110700 100644
--- a/src/reducers/versions.spec.tsx
+++ b/src/reducers/versions.spec.tsx
@@ -19,6 +19,7 @@ import reducer, {
   createInternalVersionAddon,
   createInternalVersionEntry,
   createInternalVersionFile,
+  createReducer,
   createVersionsMap,
   fetchDiff,
   fetchVersion,
@@ -143,12 +144,16 @@ describe(__filename, () => {
         { id: nextUniqueId() },
       );
 
-      let state = reducer(
+      const _reducer = createReducer({
+        _getPrunedExpandedPaths: jest.fn(getPrunedExpandedPaths),
+      })
+
+      let state = _reducer(
         undefined,
         actions.loadVersionInfo({ version: oldVersion }),
       );
-      state = reducer(state, actions.expandTree({ versionId: oldVersionId }));
-      state = reducer(state, actions.loadVersionInfo({ version: newVersion }));
+      state = _reducer(state, actions.expandTree({ versionId: oldVersionId }));
+      state = _reducer(state, actions.loadVersionInfo({ version: newVersion }));
 
       expect(state).toHaveProperty(
         'expandedPaths',
diff --git a/src/reducers/versions.tsx b/src/reducers/versions.tsx
index 616f66d..94c5233 100644
--- a/src/reducers/versions.tsx
+++ b/src/reducers/versions.tsx
@@ -1151,371 +1151,379 @@ export const goToRelativeDiff = ({
   };
 };
 
-const reducer: Reducer<VersionsState, ActionType<typeof actions>> = (
-  state = initialState,
-  action,
-): VersionsState => {
-  switch (action.type) {
-    case getType(actions.beginFetchVersion): {
-      const { versionId } = action.payload;
-
-      return {
-        ...state,
-        versionInfo: {
-          ...state.versionInfo,
-          [versionId]: undefined,
-        },
-      };
-    }
-    case getType(actions.loadVersionInfo): {
-      const { version } = action.payload;
-      const internalVersion = createInternalVersion(version);
-      const { expandedPaths: expandedPathsInState } = state;
-
-      const expandedPaths = expandedPathsInState
-        ? getPrunedExpandedPaths(expandedPathsInState, internalVersion)
-        : getParentFolders(version.file.selected_file);
-
-      return {
-        ...state,
-        expandedPaths,
-        versionInfo: {
-          ...state.versionInfo,
-          [version.id]: { ...internalVersion, expandedPaths },
-        },
-      };
-    }
-    case getType(actions.abortFetchVersion): {
-      const { versionId } = action.payload;
-
-      return {
-        ...state,
-        versionInfo: {
-          ...state.versionInfo,
-          [versionId]: null,
-        },
-      };
-    }
-    case getType(actions.loadEntryStatusMap): {
-      const { version, comparedToVersionId } = action.payload;
-      const key = getEntryStatusMapKey({
-        versionId: version.id,
-        comparedToVersionId,
-      });
-
-      return {
-        ...state,
-        entryStatusMaps: {
-          ...state.entryStatusMaps,
-          [key]: createEntryStatusMap(version),
-        },
-      };
-    }
-    case getType(actions.loadVersionFile): {
-      const { path, version } = action.payload;
-
-      return {
-        ...state,
-        versionFiles: {
-          ...state.versionFiles,
-          [version.id]: {
-            ...state.versionFiles[version.id],
-            [path]: createInternalVersionFile(version.file),
+export const createReducer = ({
+  _getPrunedExpandedPaths = getPrunedExpandedPaths,
+} = {}) => {
+  const reducer: Reducer<VersionsState, ActionType<typeof actions>> = (
+    state = initialState,
+    action,
+  ): VersionsState => {
+    switch (action.type) {
+      case getType(actions.beginFetchVersion): {
+        const { versionId } = action.payload;
+
+        return {
+          ...state,
+          versionInfo: {
+            ...state.versionInfo,
+            [versionId]: undefined,
           },
-        },
-        versionFilesLoading: {
-          ...state.versionFilesLoading,
-          [version.id]: {
-            ...state.versionFilesLoading[version.id],
-            [path]: false,
+        };
+      }
+      case getType(actions.loadVersionInfo): {
+        const { version } = action.payload;
+        const internalVersion = createInternalVersion(version);
+        const { expandedPaths: expandedPathsInState } = state;
+
+        const expandedPaths = expandedPathsInState
+          ? _getPrunedExpandedPaths(expandedPathsInState, internalVersion)
+          : getParentFolders(version.file.selected_file);
+
+        return {
+          ...state,
+          expandedPaths,
+          versionInfo: {
+            ...state.versionInfo,
+            [version.id]: { ...internalVersion, expandedPaths },
           },
-        },
-      };
-    }
-    case getType(actions.beginFetchVersionFile): {
-      const { path, versionId } = action.payload;
-
-      return {
-        ...state,
-        versionFilesLoading: {
-          ...state.versionFilesLoading,
-          [versionId]: {
-            ...state.versionFilesLoading[versionId],
-            [path]: true,
+        };
+      }
+      case getType(actions.abortFetchVersion): {
+        const { versionId } = action.payload;
+
+        return {
+          ...state,
+          versionInfo: {
+            ...state.versionInfo,
+            [versionId]: null,
           },
-        },
-      };
-    }
-    case getType(actions.abortFetchVersionFile): {
-      const { path, versionId } = action.payload;
-
-      return {
-        ...state,
-        versionFiles: {
-          ...state.versionFiles,
-          [versionId]: {
-            ...state.versionFiles[versionId],
-            [path]: null,
+        };
+      }
+      case getType(actions.loadEntryStatusMap): {
+        const { version, comparedToVersionId } = action.payload;
+        const key = getEntryStatusMapKey({
+          versionId: version.id,
+          comparedToVersionId,
+        });
+
+        return {
+          ...state,
+          entryStatusMaps: {
+            ...state.entryStatusMaps,
+            [key]: createEntryStatusMap(version),
           },
-        },
-        versionFilesLoading: {
-          ...state.versionFilesLoading,
-          [versionId]: {
-            ...state.versionFilesLoading[versionId],
-            [path]: false,
+        };
+      }
+      case getType(actions.loadVersionFile): {
+        const { path, version } = action.payload;
+
+        return {
+          ...state,
+          versionFiles: {
+            ...state.versionFiles,
+            [version.id]: {
+              ...state.versionFiles[version.id],
+              [path]: createInternalVersionFile(version.file),
+            },
           },
-        },
-      };
-    }
-    case getType(actions.updateSelectedPath): {
-      const { selectedPath, versionId } = action.payload;
-
-      const version = state.versionInfo[versionId];
-
-      if (!version) {
-        throw new Error(`Version missing for versionId: ${versionId}`);
+          versionFilesLoading: {
+            ...state.versionFilesLoading,
+            [version.id]: {
+              ...state.versionFilesLoading[version.id],
+              [path]: false,
+            },
+          },
+        };
       }
-
-      const { expandedPaths } = state;
-
-      if (!expandedPaths) {
-        throw new Error('Expanded paths are undefined');
+      case getType(actions.beginFetchVersionFile): {
+        const { path, versionId } = action.payload;
+
+        return {
+          ...state,
+          versionFilesLoading: {
+            ...state.versionFilesLoading,
+            [versionId]: {
+              ...state.versionFilesLoading[versionId],
+              [path]: true,
+            },
+          },
+        };
       }
-
-      const parents = getParentFolders(selectedPath);
-      const newExpandedPaths = [
-        ...expandedPaths,
-        ...parents.filter((newPath) => !expandedPaths.includes(newPath)),
-      ];
-
-      return {
-        ...state,
-        expandedPaths: newExpandedPaths,
-        versionInfo: {
-          ...state.versionInfo,
-          [versionId]: {
-            ...version,
-            expandedPaths: newExpandedPaths,
-            selectedPath,
+      case getType(actions.abortFetchVersionFile): {
+        const { path, versionId } = action.payload;
+
+        return {
+          ...state,
+          versionFiles: {
+            ...state.versionFiles,
+            [versionId]: {
+              ...state.versionFiles[versionId],
+              [path]: null,
+            },
           },
-        },
-      };
-    }
-    case getType(actions.setVisibleSelectedPath): {
-      const { path, versionId } = action.payload;
+          versionFilesLoading: {
+            ...state.versionFilesLoading,
+            [versionId]: {
+              ...state.versionFilesLoading[versionId],
+              [path]: false,
+            },
+          },
+        };
+      }
+      case getType(actions.updateSelectedPath): {
+        const { selectedPath, versionId } = action.payload;
 
-      const version = state.versionInfo[versionId];
+        const version = state.versionInfo[versionId];
 
-      if (!version) {
-        throw new Error(`Version missing for versionId: ${versionId}`);
-      }
+        if (!version) {
+          throw new Error(`Version missing for versionId: ${versionId}`);
+        }
 
-      if (path && !version.entries.find((e) => e.path === path)) {
-        throw new Error(
-          `Path "${path}" is an unknown path for version ID ${versionId}`,
-        );
-      }
+        const { expandedPaths } = state;
 
-      return {
-        ...state,
-        versionInfo: {
-          ...state.versionInfo,
-          [versionId]: {
-            ...version,
-            visibleSelectedPath: path,
+        if (!expandedPaths) {
+          throw new Error('Expanded paths are undefined');
+        }
+
+        const parents = getParentFolders(selectedPath);
+        const newExpandedPaths = [
+          ...expandedPaths,
+          ...parents.filter((newPath) => !expandedPaths.includes(newPath)),
+        ];
+
+        return {
+          ...state,
+          expandedPaths: newExpandedPaths,
+          versionInfo: {
+            ...state.versionInfo,
+            [versionId]: {
+              ...version,
+              expandedPaths: newExpandedPaths,
+              selectedPath,
+            },
           },
-        },
-      };
-    }
-    case getType(actions.toggleExpandedPath): {
-      const { path, versionId } = action.payload;
+        };
+      }
+      case getType(actions.setVisibleSelectedPath): {
+        const { path, versionId } = action.payload;
 
-      const version = state.versionInfo[versionId];
+        const version = state.versionInfo[versionId];
 
-      if (!version) {
-        throw new Error(`Version missing for versionId: ${versionId}`);
-      }
+        if (!version) {
+          throw new Error(`Version missing for versionId: ${versionId}`);
+        }
 
-      const { expandedPaths } = state;
+        if (path && !version.entries.find((e) => e.path === path)) {
+          throw new Error(
+            `Path "${path}" is an unknown path for version ID ${versionId}`,
+          );
+        }
 
-      if (expandedPaths === undefined) {
-        throw new Error('ExpandedPaths is undefined');
+        return {
+          ...state,
+          versionInfo: {
+            ...state.versionInfo,
+            [versionId]: {
+              ...version,
+              visibleSelectedPath: path,
+            },
+          },
+        };
       }
+      case getType(actions.toggleExpandedPath): {
+        const { path, versionId } = action.payload;
 
-      const newExpandedPaths = expandedPaths.includes(path)
-        ? expandedPaths.filter((storedPath) => path !== storedPath)
-        : [...expandedPaths, path];
+        const version = state.versionInfo[versionId];
 
-      return {
-        ...state,
-        expandedPaths: newExpandedPaths,
-        versionInfo: {
-          ...state.versionInfo,
-          [versionId]: {
-            ...version,
-            expandedPaths: newExpandedPaths,
-          },
-        },
-      };
-    }
-    case getType(actions.expandTree): {
-      const { versionId } = action.payload;
+        if (!version) {
+          throw new Error(`Version missing for versionId: ${versionId}`);
+        }
 
-      const version = state.versionInfo[versionId];
+        const { expandedPaths } = state;
+
+        if (expandedPaths === undefined) {
+          throw new Error('ExpandedPaths is undefined');
+        }
 
-      if (!version) {
-        throw new Error(`Version missing for versionId: ${versionId}`);
+        const newExpandedPaths = expandedPaths.includes(path)
+          ? expandedPaths.filter((storedPath) => path !== storedPath)
+          : [...expandedPaths, path];
+
+        return {
+          ...state,
+          expandedPaths: newExpandedPaths,
+          versionInfo: {
+            ...state.versionInfo,
+            [versionId]: {
+              ...version,
+              expandedPaths: newExpandedPaths,
+            },
+          },
+        };
       }
+      case getType(actions.expandTree): {
+        const { versionId } = action.payload;
 
-      const expandedPaths = version.entries
-        .filter((entry) => entry.type === 'directory')
-        .map((entry) => entry.path);
-      expandedPaths.push(ROOT_PATH);
+        const version = state.versionInfo[versionId];
 
-      return {
-        ...state,
-        expandedPaths,
-        versionInfo: {
-          ...state.versionInfo,
-          [versionId]: {
-            ...version,
-            expandedPaths,
+        if (!version) {
+          throw new Error(`Version missing for versionId: ${versionId}`);
+        }
+
+        const expandedPaths = version.entries
+          .filter((entry) => entry.type === 'directory')
+          .map((entry) => entry.path);
+        expandedPaths.push(ROOT_PATH);
+
+        return {
+          ...state,
+          expandedPaths,
+          versionInfo: {
+            ...state.versionInfo,
+            [versionId]: {
+              ...version,
+              expandedPaths,
+            },
           },
-        },
-      };
-    }
-    case getType(actions.collapseTree): {
-      const { versionId } = action.payload;
+        };
+      }
+      case getType(actions.collapseTree): {
+        const { versionId } = action.payload;
 
-      const version = state.versionInfo[versionId];
-      const expandedPaths = [ROOT_PATH];
+        const version = state.versionInfo[versionId];
+        const expandedPaths = [ROOT_PATH];
 
-      if (!version) {
-        throw new Error(`Version missing for versionId: ${versionId}`);
-      }
+        if (!version) {
+          throw new Error(`Version missing for versionId: ${versionId}`);
+        }
 
-      return {
-        ...state,
-        expandedPaths,
-        versionInfo: {
-          ...state.versionInfo,
-          [versionId]: {
-            ...version,
-            expandedPaths,
+        return {
+          ...state,
+          expandedPaths,
+          versionInfo: {
+            ...state.versionInfo,
+            [versionId]: {
+              ...version,
+              expandedPaths,
+            },
           },
-        },
-      };
-    }
-    case getType(actions.loadVersionsList): {
-      const { addonId, versions } = action.payload;
+        };
+      }
+      case getType(actions.loadVersionsList): {
+        const { addonId, versions } = action.payload;
+
+        return {
+          ...state,
+          byAddonId: {
+            ...state.byAddonId,
+            [addonId]: createVersionsMap(versions),
+          },
+        };
+      }
+      case getType(actions.beginFetchDiff): {
+        const key = getCompareInfoKey(action.payload);
+
+        return {
+          ...state,
+          compareInfo: {
+            ...state.compareInfo,
+            [key]: undefined,
+          },
+          compareInfoIsLoading: {
+            ...state.compareInfoIsLoading,
+            [key]: true,
+          },
+        };
+      }
+      case getType(actions.abortFetchDiff): {
+        const key = getCompareInfoKey(action.payload);
+
+        return {
+          ...state,
+          compareInfo: {
+            ...state.compareInfo,
+            [key]: null,
+          },
+          compareInfoIsLoading: {
+            ...state.compareInfoIsLoading,
+            [key]: false,
+          },
+        };
+      }
+      case getType(actions.loadDiff): {
+        const {
+          addonId,
+          baseVersionId,
+          headVersionId,
+          path,
+          version,
+        } = action.payload;
 
-      return {
-        ...state,
-        byAddonId: {
-          ...state.byAddonId,
-          [addonId]: createVersionsMap(versions),
-        },
-      };
-    }
-    case getType(actions.beginFetchDiff): {
-      const key = getCompareInfoKey(action.payload);
+        const compareInfoKey = getCompareInfoKey({
+          addonId,
+          baseVersionId,
+          headVersionId,
+          path,
+        });
 
-      return {
-        ...state,
-        compareInfo: {
-          ...state.compareInfo,
-          [key]: undefined,
-        },
-        compareInfoIsLoading: {
-          ...state.compareInfoIsLoading,
-          [key]: true,
-        },
-      };
-    }
-    case getType(actions.abortFetchDiff): {
-      const key = getCompareInfoKey(action.payload);
+        const headVersion = getVersionInfo(state, headVersionId);
+        if (!headVersion) {
+          throw new Error(
+            `Version missing for headVersionId: ${headVersionId}`,
+          );
+        }
 
-      return {
-        ...state,
-        compareInfo: {
-          ...state.compareInfo,
-          [key]: null,
-        },
-        compareInfoIsLoading: {
-          ...state.compareInfoIsLoading,
-          [key]: false,
-        },
-      };
-    }
-    case getType(actions.loadDiff): {
-      const {
-        addonId,
-        baseVersionId,
-        headVersionId,
-        path,
-        version,
-      } = action.payload;
+        const { entries, selectedPath } = headVersion;
+        const entry = entries.find((e) => e.path === selectedPath);
 
-      const compareInfoKey = getCompareInfoKey({
-        addonId,
-        baseVersionId,
-        headVersionId,
-        path,
-      });
+        if (!entry) {
+          throw new Error(`Entry missing for headVersionId: ${headVersionId}`);
+        }
 
-      const headVersion = getVersionInfo(state, headVersionId);
-      if (!headVersion) {
-        throw new Error(`Version missing for headVersionId: ${headVersionId}`);
+        return {
+          ...state,
+          compareInfo: {
+            ...state.compareInfo,
+            [compareInfoKey]: createInternalCompareInfo({
+              baseVersionId,
+              headVersionId,
+              entry,
+              version,
+            }),
+          },
+          compareInfoIsLoading: {
+            ...state.compareInfoIsLoading,
+            [compareInfoKey]: false,
+          },
+        };
       }
+      case getType(actions.setCurrentVersionId): {
+        const { versionId } = action.payload;
+        let { expandedPaths } = state;
+        const version = getVersionInfo(state, versionId);
 
-      const { entries, selectedPath } = headVersion;
-      const entry = entries.find((e) => e.path === selectedPath);
+        if (version && expandedPaths) {
+          expandedPaths = _getPrunedExpandedPaths(expandedPaths, version);
+        }
 
-      if (!entry) {
-        throw new Error(`Entry missing for headVersionId: ${headVersionId}`);
+        return {
+          ...state,
+          currentVersionId: versionId,
+          expandedPaths,
+        };
       }
-
-      return {
-        ...state,
-        compareInfo: {
-          ...state.compareInfo,
-          [compareInfoKey]: createInternalCompareInfo({
-            baseVersionId,
-            headVersionId,
-            entry,
-            version,
-          }),
-        },
-        compareInfoIsLoading: {
-          ...state.compareInfoIsLoading,
-          [compareInfoKey]: false,
-        },
-      };
-    }
-    case getType(actions.setCurrentVersionId): {
-      const { versionId } = action.payload;
-      let { expandedPaths } = state;
-      const version = getVersionInfo(state, versionId);
-
-      if (version && expandedPaths) {
-        expandedPaths = getPrunedExpandedPaths(expandedPaths, version);
+      case getType(actions.unsetCurrentVersionId): {
+        return {
+          ...state,
+          currentVersionId: false,
+        };
       }
-
-      return {
-        ...state,
-        currentVersionId: versionId,
-        expandedPaths,
-      };
-    }
-    case getType(actions.unsetCurrentVersionId): {
-      return {
-        ...state,
-        currentVersionId: false,
-      };
+      default:
+        return state;
     }
-    default:
-      return state;
-  }
+  };
+
+  return reducer;
 };
 
-export default reducer;
+export default createReducer();
```

</details>